Allows preprocess_weights_for_loading() to consider layers wrapped in TimeDistributed or Bidirectional

Russ09 opened the pull request:
Currently layers with a Bidirectional or TimeDistributed wrapper are not considered by preprocess_weights_for_loading() as only the layer.__class__.__name__ property is examined. This would fix that. Alternatively it could be addressed within preprocess_weights_for_loading()

fchollet commented:
Please add unit tests.

Russ09 commented:
Since this mainly concerns reading weight files from Keras 1, I am not 100% sure what can helpfully be covered in the Keras 2 unit tests, any thoughts? I also didn't see any unit tests for the general reading and writing of weights, is it permissible to write files within the unit tests?

fchollet commented:
Currently, how was correctness tested? Do you have a test script and test files that you used?

fchollet commented:
Any update?

Russ09 commented:
Sorry, PhD work built up and then I totally forgot about this! So I had tested it for myself by having a weight file (.hdf5) saved from an older version and then loading it in and testing that the weights in each layer had been modified, but that requires lugging around an old weights file. Perhaps it would make more sense to move the fix into preprocess_weights_for_loading() then unit tests could just work with weight tensors that are constructed in the old style?

fchollet commented:
Agreed, `preprocess_weights_for_loading` is where this step belongs.

One note on style: use e.g. `num_weights` instead of `nb_weights`.

Russ09 commented:
I have refactored `preprocess_weights_for_loading` to handle the cases from TimeDistributed and Bidirectional wrappers. I have added unit tests that construct a TimeDistributed and a Bidirectional wrapped layer and construct weight tensors that need preprocessing via `preprocess_weights_for_loading`. I assert that the input tensors and the layer weights are equal after preprocessing and batch applying them.

fchollet commented on keras/engine/topology.py:
```diff
@@ -2746,6 +2746,28 @@ def preprocess_weights_for_loading(layer, weights,
+    return weights
+
+
+def _preprocess_weights(layer, weights, original_keras_version=None, original_backend=None):
```
Why do you need two methods? You should only need one...

Russ09 replied:
I felt like it made more sense to do separate calls for `layer.forward_layer` and `layer.backward_layer` in the bidirectional case since then you are passing an actual instance of the correct child layer into `_preprocess_weights`. You could do it in a single method but it would involve either duplicating code or specifically tailoring parts of `_preprocess_weights` to  cater for the bidirectional case where there are twice as many weight tensors, and the list of weight tensors is hard-coded with indices in that method already.

fchollet replied:
Why not recursively call `preprocess_weights_for_loading` in the bidirectional case inside `preprocess_weights_for_loading`?

Russ09 replied:
A fair point, that would work fine. I will commit that tomorrow morning as I am on UK time!

fchollet commented on keras/engine/topology.py:
```diff
@@ -2956,11 +2976,12 @@ def load_weights_from_hdf5_group_by_name(f, layers):
+                                                           weight_values,
+                                                           original_keras_version,
+                                                           original_backend)
+
```
Why this whitespace change?

fchollet commented on keras/engine/topology.py:
```diff
@@ -2746,102 +2746,120 @@ def preprocess_weights_for_loading(layer, weights,
+                                                          original_keras_version,
+                                                          original_backend)
+        weights = forward_weights + backward_weights
+    elif layer.__class__.__name__ == 'TimeDistributed':
```
These two entries (if Bidirectional, if TimeDistributed) should be added in the big `if` switch without affecting the rest of the code. Currently we have 150 lines changed for should be a ~15 lines change...

Russ09 replied:
I am not sure I follow, I currently have it as:

`if TimeDistributed:
    #recurse
elif Bidirectional:
    #recurse
else:
    #original code` 

Would this be cleaner as a new PR? Because I think a lot of the changes are essentially reverting my original suggestion.

Russ09 replied:
Sorry, I see what you meant. Think I have addressed that now.

fchollet commented on keras/engine/topology.py:
```diff
@@ -2832,26 +2848,24 @@ def preprocess_weights_for_loading(layer, weights,
+                weights[0] = conv_utils.convert_kernel(weights[0])
+            if layer.__class__.__name__ == 'ConvLSTM2D':
+                weights[0] = conv_utils.convert_kernel(weights[0])
+                weights[1] = conv_utils.convert_kernel(weights[1])
```
I don't understand the reindentation here. The initial indentation was the correct one.

fchollet commented on keras/engine/topology.py:
```diff
@@ -2915,15 +2929,11 @@ def load_weights_from_hdf5_group(f, layers):
     # Arguments
         f: A pointer to a HDF5 group.
         layers: a list of target layers.
-
```
All 6 deleted newlines above should be kept (docstring formatting conventions).

fchollet commented on keras/engine/topology.py:
```diff
@@ -2747,6 +2747,22 @@ def preprocess_weights_for_loading(layer, weights,
+            weights = forward_weights + backward_weights
+
+        if layer.__class__.__name__ == 'TimeDistributed':
+            weights = preprocess_weights_for_loading(layer.layer, weights, original_keras_version, original_backend)
```
Order arguments vertically to avoid an overly long line.

fchollet commented:
LGTM, thanks

Russ09 commented:
great, should this be submitted to tf.contrib.keras also?
